### PR TITLE
rustc/rustc_mir: Implement RFC 2203.

### PR DESCRIPTION
**Pull request by davidtwco**
This PR implements RFC 2203, allowing constants in array repeat
expressions. Part of #49147.

r? @eddyb

### CODE REVIEW

File: src/librustc/mir/mod.rs [hunk 1]
```diff
@@ -2096,6 +2096,15 @@ impl<'tcx> Place<'tcx> {
             Place::Base(base) => op(base, next.iter()),
         }
     }
+
+    /// Returns `true` if this is a promoted place.
+    pub fn is_promoted(&self) -> bool {
```
Comment (eddyb) [inline src/librustc/mir/mod.rs]: I wouldn't add helpers for a temporary hack.

File: src/test/ui/consts/rfc-2203-const-array-repeat-exprs/migrate-borrowck.rs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+    fn impl_copy_value_one_element() {
+        const FOO: Option<u32> = Some(4);
+        const ARR: [Option<u32>; 1] = [FOO; 1];
+    }
+
+    fn impl_copy_value_multiple_elements() {
+        const FOO: Option<u32> = Some(4);
+        const ARR: [Option<u32>; 2] = [FOO; 2];
+    }
```
Comment (Centril) [inline src/test/ui/consts/rfc-2203-const-array-repeat-exprs/migrate-borrowck.rs]: I was going to say we should add a test where the repeat amount is unknown (a generic value from an associated constant and/or const generics) but iirc the repeat amount is an AnonConst so we cannot?

File: src/test/ui/consts/rfc-2203-const-array-repeat-exprs/migrate-borrowck.rs [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        let arr: [Option<u32>; 0] = [x; 0];
+    }
+
+    fn impl_copy_empty_value_one_element() {
+        let x: Option<u32> = None;
+        let arr: [Option<u32>; 1] = [x; 1];
+    }
+
+    fn impl_copy_empty_value_multiple_elements() {
+        let x: Option<u32> = None;
```
Comment (Centril) [inline src/test/ui/consts/rfc-2203-const-array-repeat-exprs/migrate-borrowck.rs]: o.O; This is allowing non-constants as well?

File: src/test/ui/consts/rfc-2203-const-array-repeat-exprs/migrate-borrowck.rs [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+    fn no_impl_copy_empty_value_single_element() {
+        let x = None;
+        let arr: [Option<Bar>; 1] = [x; 1];
+    }
+
+    fn no_impl_copy_empty_value_multiple_elements() {
+        let x = None;
+        let arr: [Option<Bar>; 2] = [x; 2];
+        //~^ ERROR repeated expression does not implement `std::marker::Copy`
```
Comment (Centril) [inline src/test/ui/consts/rfc-2203-const-array-repeat-exprs/migrate-borrowck.rs]: Please separate out the bits that are supposed to pass vs. not pass from each other into different files (and in the next file as well).

Would also be good to have some run-pass tests that actually exercises the dynamic behavior.

File: src/librustc_mir/borrow_check/nll/type_check/mod.rs [hunk 5]
```diff
@@ -1947,18 +1949,32 @@ impl<'a, 'gcx, 'tcx> TypeChecker<'a, 'gcx, 'tcx> {
...
-                    location.to_locations(),
-                    ConstraintCategory::CopyBound,
-                );
+                if let Operand::Move(_) = operand {
+                    // While this is located in `nll::typeck` this error is not an NLL error, it's
+                    // a required check to make sure that repeated elements implement `Copy`.
+                    let span = mir.source_info(location).span;
+                    let ty = operand.ty(mir, self.tcx());
+                    let is_copy = self.infcx.type_is_copy_modulo_regions(self.param_env, ty, span);
+                    if !is_copy {
```
Comment (Centril) [inline src/librustc_mir/borrow_check/nll/type_check/mod.rs]: Can you move the diagnostics logic to a separate function to keep the happy path cleaner?

File: src/librustc_mir/transform/promote_consts.rs [hunk 6]
```diff
@@ -322,6 +325,17 @@ impl<'a, 'tcx> Promoter<'a, 'tcx> {
...
                 }
+                Candidate::Repeat(loc) => {
+                    let ref mut statement = blocks[loc.block].statements[loc.statement_index];
+                    match statement.kind {
+                        StatementKind::Assign(_, box Rvalue::Repeat(ref mut operand, _)) => {
+                            let ty = operand.ty(local_decls, self.tcx);
+                            let span = statement.source_info.span;
+                            mem::replace(operand, Operand::Copy(promoted_place(ty, span)))
+                        }
+                        _ => bug!()
```
Comment (Centril) [inline src/librustc_mir/transform/promote_consts.rs]:
```suggestion
                        _ => bug!("something something fill me in")
```

File: src/librustc_typeck/check/mod.rs [hunk 7]
```diff
@@ -4453,16 +4453,6 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                     }
                 };
 
-                if let Ok(count) = count {
-                    let zero_or_one = count.assert_usize(tcx).map_or(false, |count| count <= 1);
-                    if !zero_or_one {
```
Comment (Centril) [inline src/librustc_typeck/check/mod.rs]: Presumably we'll want to feature gate this?

File: src/librustc/mir/mod.rs (hunk 1 above)
Comment (davidtwco) [reply src/librustc/mir/mod.rs]: Fixed.

File: src/librustc_mir/transform/promote_consts.rs (hunk 6 above)
Comment (eddyb) [reply src/librustc_mir/transform/promote_consts.rs]: This is pre-existing, and it's a shorter `unreachable!`, in that it should be impossible to hit this (i.e. the `Candidate` is bogus).

File: src/librustc_typeck/check/mod.rs (hunk 7 above)
Comment (eddyb) [reply src/librustc_typeck/check/mod.rs]: We can feature-gate the promotion.

File: src/test/ui/consts/rfc-2203-const-array-repeat-exprs/migrate-borrowck.rs (hunk 2 above)
Comment (eddyb) [reply src/test/ui/consts/rfc-2203-const-array-repeat-exprs/migrate-borrowck.rs]: `AnonConst` is not the problem, but this is an instance of "knowledge of generics in scopes cannot reach constants because without lazy normalization lots of things just become cycle errors". There's been an issue open for years but I can never remember how to find it.

Comment (eddyb) [reply src/test/ui/consts/rfc-2203-const-array-repeat-exprs/migrate-borrowck.rs]: Might work with const generics though?

File: src/test/ui/consts/rfc-2203-const-array-repeat-exprs/migrate-borrowck.rs (hunk 3 above)
Comment (eddyb) [reply src/test/ui/consts/rfc-2203-const-array-repeat-exprs/migrate-borrowck.rs]: No, this wouldn't get promoted, `Option<u32>` just happens to be `Copy`.

File: src/librustc_mir/borrow_check/nll/type_check/mod.rs [hunk 8]
```diff
@@ -1947,18 +1957,32 @@ impl<'a, 'gcx, 'tcx> TypeChecker<'a, 'gcx, 'tcx> {
...
+                            self.tcx().lang_items().copy_trait().unwrap());
+                        self.tcx().sess
+                            .struct_span_err(
+                                span,
+                                &format!("repeated expression does not implement `{}`", copy_path),
+                            )
+                            .span_label(span, &format!(
+                                "the trait `{}` is not implemented for `{}`",
+                                copy_path, ty,
+                            ))
```
Comment (eddyb) [inline src/librustc_mir/borrow_check/nll/type_check/mod.rs]: cc @nikomatsakis @matthewjasper This is not a proper trait error, so it would miss the "backtrace" that one might get otherwise - any suggestions on how to fix that?

**Comment by eddyb** [comment]
To expand on https://github.com/rust-lang/rust/pull/61749#discussion_r292756424, this example:
```rust
#[derive(Copy, Clone)]
struct Foo<T>(T);
fn main() {
    [Foo(String::new()); 4];
}
```
Currently errors with:
```rust
error[E0277]: the trait bound `std::string::String: std::marker::Copy` is not satisfied
 --> src/main.rs:4:5
  |
4 |     [Foo(String::new()); 4];
  |     ^^^^^^^^^^^^^^^^^^^^^^^ the trait `std::marker::Copy` is not implemented for `std::string::String`
  |
  = note: required because of the requirements on the impl of `std::marker::Copy` for `Foo<std::string::String>`
  = note: the `Copy` trait is required because the repeated element will be copied
```
The "required because of the requirements on the impl of ..." part would now be missing, I think.
There might be a relatively easy way to produce a proper trait error, even from the NLL `type_check`.
Either way, it would be good to have tests like this to properly compare the effect.

File: src/test/ui/consts/rfc-2203-const-array-repeat-exprs/migrate-borrowck.rs (hunk 2 above)
Comment (davidtwco) [reply src/test/ui/consts/rfc-2203-const-array-repeat-exprs/migrate-borrowck.rs]: I don't think this would work const generics either unfortunately (see [this test case from a recent PR of mine](https://github.com/rust-lang/rust/pull/61698/files#diff-493e70c94f4001daf2645ddf3db803b5R10)).

File: src/test/ui/consts/rfc-2203-const-array-repeat-exprs/migrate-borrowck.rs (hunk 4 above)
Comment (davidtwco) [reply src/test/ui/consts/rfc-2203-const-array-repeat-exprs/migrate-borrowck.rs]: I've separated the test files. Unfortunately, I only have time to add a very simple run-pass test at the moment.

File: src/librustc_mir/borrow_check/nll/type_check/mod.rs (hunk 5 above)
Comment (davidtwco) [reply src/librustc_mir/borrow_check/nll/type_check/mod.rs]: I've started re-using some existing functions for diagnostics here.

File: src/librustc_typeck/check/mod.rs (hunk 7 above)
Comment (davidtwco) [reply src/librustc_typeck/check/mod.rs]: I've added a feature gate and use it in `qualif_consts` (I'm not sure if you wanted it to be used in `promote_consts` or `qualif_consts` but it was slightly easier add to `qualif_consts`).

File: src/librustc_mir/borrow_check/nll/type_check/mod.rs (hunk 8 above)
Comment (davidtwco) [reply src/librustc_mir/borrow_check/nll/type_check/mod.rs]: I've started re-using some existing functions for diagnostics in this PR. It appears to work correctly but it might not produce the full "backtrace" that you want.

**Comment by davidtwco** [comment]
> Either way, it would be good to have tests like this to properly compare the effect.

I've added this example as a test.

**Comment by eddyb** [comment]
@davidtwco Sadly, without adding a test like this (or a more complex one) on master first, it's hard to see the changes in diagnostics.

**Comment by davidtwco** [comment]
@eddyb I've ran that example with another working directory (at most 24 hours old behind master) and this is the current error:
```
error[E0277]: the trait bound `std::string::String: std::marker::Copy` is not satisfied
 --> ../rust0/src/test/ui/consts/rfc-2203-const-array-repeat-exprs/trait-error.rs:8:5
  |
8 |     [Foo(String::new()); 4];
  |     ^^^^^^^^^^^^^^^^^^^^^^^ the trait `std::marker::Copy` is not implemented for `std::string::String`
  |
  = note: required because of the requirements on the impl of `std::marker::Copy` for `Foo<std::string::String>`
  = note: the `Copy` trait is required because the repeated element will be copied

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
```

File: src/test/ui/consts/rfc-2203-const-array-repeat-exprs/migrate-borrowck.rs (hunk 3 above)
Comment (Centril) [reply src/test/ui/consts/rfc-2203-const-array-repeat-exprs/migrate-borrowck.rs]: Oh... I read too fast...

File: src/libsyntax/feature_gate.rs [hunk 9]
```diff
@@ -570,6 +570,9 @@ declare_features! (
     // #[repr(transparent)] on unions.
     (active, transparent_unions, "1.37.0", Some(60405), None),
 
+    // `[x; N]` where `x` is a constant (RFC 2203).
```
Comment (Centril) [inline src/libsyntax/feature_gate.rs]:
```suggestion
    // Allows `[x; N]` where `x` is a constant (RFC 2203).
```

Can you also drive-by fix the two feature gates above that don't conform to the style?

File: src/librustc_mir/transform/qualify_consts.rs [hunk 10]
```diff
@@ -935,6 +948,16 @@ impl<'a, 'tcx> Checker<'a, 'tcx> {
         debug!("qualify_const: promotion_candidates={:?}", self.promotion_candidates);
         for candidate in &self.promotion_candidates {
             match *candidate {
+                Candidate::Repeat(Location { block: bb, statement_index: stmt_idx }) => {
+                    match self.body[bb].statements[stmt_idx].kind {
```
Comment (Centril) [inline src/librustc_mir/transform/qualify_consts.rs]: Use `if let`?

File: src/test/ui/consts/rfc-2203-const-array-repeat-exprs/migrate-pass.rs [hunk 11]
```diff
@@ -0,0 +1,128 @@
+// compile-pass
+// compile-flags: -Z borrowck=migrate
+// ignore-compare-mode-nll
+#![feature(const_in_array_repeat_expressions)]
+#![allow(warnings)]
+
+// Some type that is not copyable.
+struct Bar;
+
+mod constants {
```
[the remark this answers is not in the document]
Comment (davidtwco) [reply src/test/ui/consts/rfc-2203-const-array-repeat-exprs/migrate-pass.rs]: I've pushed a commit that adds some tests using `const fn`.

File: src/test/ui/consts/rfc-2203-const-array-repeat-exprs/migrate-const-fn-pass.rs [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+#![feature(const_in_array_repeat_expressions)]
+#![allow(warnings)]
+
+// Some type that is not copyable.
+struct Bar;
+
+mod constants {
+    use Bar;
+
+    const fn no_impl_copy_empty_value_no_elements() {
```
Comment (Centril) [inline src/test/ui/consts/rfc-2203-const-array-repeat-exprs/migrate-const-fn-pass.rs]: Oh... hmm; I suppose I should have been clearer... =)

What I meant is calling a `const fn` for the element expression.

Comment (davidtwco) [reply src/test/ui/consts/rfc-2203-const-array-repeat-exprs/migrate-const-fn-pass.rs]: Fixed.

**Comment by eddyb** [comment]
What is this blocked on, if anything?

**Comment by davidtwco** [comment]
I'm not aware of any outstanding concerns, only that it isn't a proper trait error with a backtrace being emitted.

**Comment by eddyb** [comment]
@davidtwco I'm sorry for blocking this, I don't think the diagnostic regressions are worth the wait.
Especially given situations like https://github.com/rust-lang/rust/issues/53491#issuecomment-512822519.

File: src/test/ui/const-generics/issue-61336-2.rs [hunk 13]
```diff
@@ -3,11 +3,12 @@
 
 fn f<T: Copy, const N: usize>(x: T) -> [T; N] {
     [x; {N}]
+    //~^ ERROR array lengths can't depend on generic parameters
```
Comment (eddyb) [inline src/test/ui/const-generics/issue-61336-2.rs]: Huh I don't remember this error.

Comment (eddyb) [reply src/test/ui/const-generics/issue-61336-2.rs]: Oh it's for `[x; n]` expressions, not types. Makes sense (although it should use a `ty::Const` instead of hardcoding an integer, cc @varkor @oli-obk ).

**Review by eddyb — APPROVED** [review]

File: src/test/ui/feature-gates/feature-gate-const_in_array_repeat_expressions.stderr [hunk 14]
```diff
@@ -0,0 +1,13 @@
+error[E0277]: the trait bound `std::option::Option<std::string::String>: std::marker::Copy` is not satisfied
+  --> $DIR/feature-gate-const_in_array_repeat_expressions.rs:7:36
+   |
+LL |     let arr: [Option<String>; 2] = [None::<String>; 2];
+   |                                    ^^^^^^^^^^^^^^^^^^^ the trait `std::marker::Copy` is not implemented for `std::option::Option<std::string::String>`
+   |
+   = help: the following implementations were found:
+             <std::option::Option<T> as std::marker::Copy>
+   = note: the `Copy` trait is required because the repeated element will be copied
```
Comment (Centril) [inline src/test/ui/feature-gates/feature-gate-const_in_array_repeat_expressions.stderr]: This doesn't mention the feature gate :(

File: src/librustc_mir/transform/qualify_consts.rs [hunk 15]
```diff
@@ -933,15 +946,20 @@ impl<'a, 'tcx> Checker<'a, 'tcx> {
...
         for candidate in &self.promotion_candidates {
             match *candidate {
+                Candidate::Repeat(Location { block: bb, statement_index: stmt_idx }) => {
+                    if let StatementKind::Assign(_, box Rvalue::Repeat(
+                        Operand::Move(Place::Base(PlaceBase::Local(index))),
+                        _
+                    )) = self.body[bb].statements[stmt_idx].kind {
+                        promoted_temps.insert(index);
+                    }
+                }
```
Comment (eddyb) [inline src/librustc_mir/transform/qualify_consts.rs]: I just realized, this is unnecessary, only `Candidate::Ref` cares (it removes `StorageDead` from the temp that would need to be `'static`).

File: src/librustc_mir/transform/qualify_consts.rs [hunk 16]
```diff
@@ -933,15 +946,20 @@ impl<'a, 'tcx> Checker<'a, 'tcx> {
...
-                            box Rvalue::Ref(_, _, Place::Base(PlaceBase::Local(index)))
-                        ) => {
-                            promoted_temps.insert(index);
-                        }
-                        _ => {}
+                    if let StatementKind::Assign(
+                        _,
+                        box Rvalue::Ref(_, _, Place::Base(PlaceBase::Local(index)))
+                    ) = self.body[bb].statements[stmt_idx].kind {
+                        promoted_temps.insert(index);
```
Comment (eddyb) [inline src/librustc_mir/transform/qualify_consts.rs]: @oli-obk Wow this is wrong?! `const FOO: &i32 = { let x = &(5, false).0; x };` fails because of it, lol.

File: src/test/ui/feature-gates/feature-gate-const_in_array_repeat_expressions.rs [hunk 17]
```diff
@@ -0,0 +1,11 @@
+// ignore-tidy-linelength
+#![allow(warnings)]
+
+struct Bar;
+
+fn foo() {
+    let arr: [Option<String>; 2] = [None::<String>; 2];
+    //~^ ERROR the trait bound `std::option::Option<std::string::String>: std::marker::Copy` is not satisfied [E0277]
+}
+
```
Comment (RalfJung) [inline src/test/ui/feature-gates/feature-gate-const_in_array_repeat_expressions.rs:10]: Turns out this test was not thorough enough, and [some other cases sneaked past the feature gate](https://github.com/rust-lang/rust/issues/49147#issuecomment-726796665).
